### PR TITLE
special characters during project name generation 

### PR DESCRIPTION
**Pull request by sharabanys**
### Description

Add new parameter **preserve-project-name**  in order to allow a choice whether to leave the name of the project that will open in the review in the name of the REPO or not.
When **False**: The project name for the scanner will be the repository name after normalization. Legal characters are: 'a-z', 'A-Z', '0-9', "-", "_", ".". All other characters will be replaced in the normalization process with "-". 
When **True**: The project name for the scanner will be the same as the repository name without normalization.
In case the customer created a project directly in scanner under a non-normalized name, the "True" check will give him access to the same parquet, otherwise he will create a new project in scanner.
Please note: Not all scanners allow project names with invalid characters.

### References

https://checkmarx.atlassian.net/browse/AST-1326
link to updated WIKI: https://github.com/checkmarx-ltd/cx-flow/wiki/Configuration#configuration-definitions

### Testing
Add project-name-generator.feature 

### Checklist

- [X] I have added documentation for new functionality in this PR (if applicable).  *If documentation is a Wiki Update, please indicate desired changes within PR MD Comment*
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [X] The correct base branch is being used

### CODE REVIEW

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
Logo<br />**Checkmarx SAST - Scan Summary & Details**
### Cx-SAST Summary
Total of 1412 vulnerabilities<br />High&nbsp;**0 High**<br />Medium&nbsp;**44 Medium**<br />Low&nbsp;**1262 Low**<br />Info&nbsp;**106 Info**<br />
### Violation Summary
Medium&nbsp;**10** **Medium**<br />
[View more details on Checkmarx UI](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1058670&projectid=1951)
### Cx-SAST Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---
[67](https://github.com/checkmarx-ltd/cx-flow/blob/pr-saras-project-name-generation/src/main/java/com/checkmarx/flow/controller/FlowController.java#L67) [70](https://github.com/checkmarx-ltd/cx-flow/blob/pr-saras-project-name-generation/src/main/java/com/checkmarx/flow/controller/FlowController.java#L70) [71](https://github.com/checkmarx-ltd/cx-flow/blob/pr-saras-project-name-generation/src/main/java/com/checkmarx/flow/controller/FlowController.java#L71) [76](https://github.com/checkmarx-ltd/cx-flow/blob/pr-saras-project-name-generation/src/main/java/com/checkmarx/flow/controller/FlowController.java#L76) [78](https://github.com/checkmarx-ltd/cx-flow/blob/pr-saras-project-name-generation/src/main/java/com/checkmarx/flow/controller/FlowController.java#L78) [137](https://github.com/checkmarx-ltd/cx-flow/blob/pr-saras-project-name-generation/src/main/java/com/checkmarx/flow/controller/FlowController.java#L137) |Medium|Input_Path_Not_Canonicalized|src/main/java/com/checkmarx/flow/controller/FlowController.java|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1058670&projectid=1951&pathid=23)
[100](https://github.com/checkmarx-ltd/cx-flow/blob/pr-saras-project-name-generation/src/main/java/com/checkmarx/flow/controller/GitHubController.java#L100) [242](https://github.com/checkmarx-ltd/cx-flow/blob/pr-saras-project-name-generation/src/main/java/com/checkmarx/flow/controller/GitHubController.java#L242) |Medium|Input_Path_Not_Canonicalized|src/main/java/com/checkmarx/flow/controller/GitHubController.java|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1058670&projectid=1951&pathid=11)
[60](https://github.com/checkmarx-ltd/cx-flow/blob/pr-saras-project-name-generation/src/main/java/com/checkmarx/flow/service/ExternalScriptService.java#L60) |Medium|Input_Path_Not_Canonicalized|src/main/java/com/checkmarx/flow/service/ExternalScriptService.java|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1058670&projectid=1951&pathid=50)
[213](https://github.com/checkmarx-ltd/cx-flow/blob/pr-saras-project-name-generation/src/main/java/com/checkmarx/flow/utils/ScanUtils.java#L213) |Medium|Input_Path_Not_Canonicalized|src/main/java/com/checkmarx/flow/utils/ScanUtils.java|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1058670&projectid=1951&pathid=7)

File: src/test/resources/cucumber/features/componentTests/project-name-generator.feature [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    Given is-multi-tenant <is-multi-tenant>
+    When scan request arrives with namespace <namespace>, repo-name <repo-name>, branch <branch> and application <application>
+    Then project name used by scanner is <project-name>
+
+    Examples:
+      | namespace     | repo-name | branch | application           | is-multi-tenant | project-name                 |
+      | checkmarx-ltd | cx-flow   |        | N/A                   | true            | cx-flow                      |
+      | checkmarx-ltd | cx-flow   | master | N/A                   | true            | cx-flow-master               |
+      | checkmarx-ltd | cx-flow   | master | N/A                   | false           | checkmarx-ltd-cx-flow-master |
+      |               |           |        | checkmarx-application | false           | checkmarx-application        |
```
Comment (alex-ko-dev) [inline src/test/resources/cucumber/features/componentTests/project-name-generator.feature:23]: Maybe we need more cases here? E.g. the same params as in the last row but with multi-tenant=true.

File: docs/Configuration.md [hunk 2]
```diff
@@ -339,6 +340,8 @@ Refer to the sample configuration above for the entire yaml structure.
 | profiling-depth   | 1             | No  | Yes | No        | The folder depth that is inspected for file names during the profiling process, which means looking for specific file references, i.e. web.xml/Web.config |
 | profile-config    | CxProfile.json | No | Yes | No        | The file that contains the profile configuration mapping. |
 | scan-resubmit     | false          | No | Yes | Yes       | When **True**: If a scan is active for the same project, CxFlow cancels the active scan and submits a new scan. When **False**: If a scan is active for the same project, CxFlow does not submit a new scan. |
+| preserve-project-name | false      | No | Yes | Yes       | When **False**: The project name for the scanner will be the repository name after normalization. Legal characters are: 'a-z', 'A-Z', '0-9', "-", "_", ".". All other characters will be replaced in the normalization process with "-".
```
Comment (sharabanys) [inline docs/Configuration.md:343]: @jbrotsos 
Could you review the text?

File: src/main/java/com/checkmarx/flow/service/ProjectNameGenerator.java [hunk 3]
```diff
@@ -59,16 +62,21 @@ public String determineProjectName(ScanRequest request) {
...
-                log.debug("Project name ({}) has been normalized to allow only valid characters.", rawProjectName);
+            if (!preserveProjectName) {
+                //only allow specific chars in project name in checkmarx
+                result = rawProjectName.replaceAll("[^a-zA-Z0-9-_.]+", "-");
+                if (!result.equals(rawProjectName)) {
+                    log.debug("Project name ({}) has been normalized to allow only valid characters.", rawProjectName);
+                }
+            } else {
+                result = rawProjectName;
+                log.info("Project name ({}) has not been normalized.", rawProjectName);
```
Comment (sharabanys) [inline src/main/java/com/checkmarx/flow/service/ProjectNameGenerator.java:79]: @AvivCx 
Could you review this log?

File: src/main/java/com/checkmarx/flow/config/FlowProperties.java [hunk 4]
```diff
@@ -47,6 +47,7 @@
     private boolean listFalsePositives = false;
     private boolean scanResubmit = false;
     private Mail mail;
+    private boolean preserveProjectName = false;
```
[the remark this answers is not in the document]
Comment (sharabanys) [reply src/main/java/com/checkmarx/flow/config/FlowProperties.java:50]: thanks

File: src/test/java/com/checkmarx/flow/cucumber/component/projectname/ProjectNameSteps.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    private final HelperService helperService;
+
+    public ProjectNameSteps(CxScannerService cxScannerService, FlowProperties flowProperties, ProjectNameGenerator projectNameGenerator, HelperService helperService) {
+        this.flowProperties = flowProperties;
+        this.cxScannerService = cxScannerService;
+        this.projectNameGenerator = projectNameGenerator;
+        this.helperService = mock(HelperService.class);
+    }
+
+    @Given("preserve-project-name is {} and is-multi-tenant true")
```
[the remark this answers is not in the document]
Comment (sharabanys) [reply src/test/java/com/checkmarx/flow/cucumber/component/projectname/ProjectNameSteps.java:36]: thanks

File: src/test/resources/cucumber/features/componentTests/project-name-generator.feature (hunk 1 above)
Comment (sharabanys) [reply src/test/resources/cucumber/features/componentTests/project-name-generator.feature:23]: The cases is according to code,
I did't write all options

File: docs/Configuration.md [hunk 6]
```diff
@@ -339,6 +340,7 @@ Refer to the sample configuration above for the entire yaml structure.
 | profiling-depth   | 1             | No  | Yes | No        | The folder depth that is inspected for file names during the profiling process, which means looking for specific file references, i.e. web.xml/Web.config |
 | profile-config    | CxProfile.json | No | Yes | No        | The file that contains the profile configuration mapping. |
 | scan-resubmit     | false          | No | Yes | Yes       | When **True**: If a scan is active for the same project, CxFlow cancels the active scan and submits a new scan. When **False**: If a scan is active for the same project, CxFlow does not submit a new scan. |
+| preserve-project-name | false      | No | Yes | Yes       | When **False**: The project name for the scanner will be the repository name after normalization. Legal characters are: 'a-z', 'A-Z', '0-9', "-", "_", ".". All other characters will be replaced in the normalization process with "-". When **True**: The project name for the scanner will be the same as the repository name without normalization. For attention: Not all scanners allow project names with invalid characters. |
```
[the remark this answers is not in the document]
Comment (sharabanys) [reply docs/Configuration.md]: Thanks

File: docs/Configuration.md (hunk 2 above)
Comment (jbrotsos) [reply docs/Configuration.md:343]: My suggestion:

When **False**: The project name will be the repository name followed by a dash followed by the branch name after normalization (i.e. front-end-dev). Legal characters are: 'a-z', 'A-Z', '0-9', "-", "_", ".". All other characters will be replaced in the normalization process with "-".

File: docs/Configuration.md [hunk 7]
```diff
@@ -339,6 +340,8 @@ Refer to the sample configuration above for the entire yaml structure.
 | profiling-depth   | 1             | No  | Yes | No        | The folder depth that is inspected for file names during the profiling process, which means looking for specific file references, i.e. web.xml/Web.config |
 | profile-config    | CxProfile.json | No | Yes | No        | The file that contains the profile configuration mapping. |
 | scan-resubmit     | false          | No | Yes | Yes       | When **True**: If a scan is active for the same project, CxFlow cancels the active scan and submits a new scan. When **False**: If a scan is active for the same project, CxFlow does not submit a new scan. |
+| preserve-project-name | false      | No | Yes | Yes       | When **False**: The project name for the scanner will be the repository name after normalization. Legal characters are: 'a-z', 'A-Z', '0-9', "-", "_", ".". All other characters will be replaced in the normalization process with "-".
+When **True**: The project name for the scanner will be the same as the repository name without normalization. For attention: Not all scanners allow project names with invalid characters. |
```
Comment (jbrotsos) [inline docs/Configuration.md:344]: When **True**: The project name for the scanner will be the exact project name inputted without normalization (i.e. front end). For attention: Not all scanners allow project names with invalid characters.

**Review by alex-ko-dev — APPROVED** [review]

**Review by alex-ko-dev — APPROVED** [review]

**Review by alex-ko-dev — APPROVED** [review]

**Pull request merged**